cpp/src/arrow/scalar.cc

```diff
@@ -542,6 +542,21 @@ struct ScalarValidateImpl {
   }
 };
 
+template <size_t offset, typename T, typename... Args>
+void FillScalarScratchSpaceHelper(uint8_t* scratch_space, T first, Args... args) {
+  static_assert(offset + sizeof(T) <= internal::kScalarScratchSpaceSize);
+  *reinterpret_cast<T*>(scratch_space + offset) = first;
+  if constexpr (sizeof...(args) > 0) {
+    FillScalarScratchSpaceHelper<offset + sizeof(T)>(scratch_space,
+                                                     std::forward<Args>(args)...);
+  }
+}
+
+template <typename... Args>
+void FillScalarScratchSpace(Args... args) {
+  FillScalarScratchSpaceHelper<0>(std::forward<Args>(args)...);
+}
+
```

Comment:
I think this could be simplified to:
```c++
template <typename T, size_t N>
void WriteArray(void *d, T const (&arr)[N]) {
  static_assert(sizeof(arr) <= internal::kScalarScratchSpaceSize);
  std::memcpy(d, arr, sizeof(arr));
}

void BinaryScalar::FillScratchSpace() {
  WriteArray(scratch_space_,
             {int32_t(0), value ? static_cast<int32_t>(value->size()) : int32_t(0)});
}
```

Reply:
Oh, I wanted to make `FillScalarScratchSpace` able to accept arbitrary number of arguments of arbitrary types, meanwhile do the copy type-safely.

Reply:
`FillScalarScratchSpaceHelper` doesn't need to write arbitrary arguments, and actually there's a risk of undefined behavior as written since you might `reinterpret_cast<int32_t*>(scratch_space + offset/*=1*/)`

Reply:
I understand that it is unnecessary for this function to accept arbitrary arguments. I will change it for the sake of that. 

However I am curious about the risk of UB you mentioned. Did you mean C++ UB wrt. strong aliasing rule? If yes, will memcpy instead eliminate this kind of UB?

Reply:
Addressed. But my curiosity remains :)

Thank you.

Reply:
The UB I'm referring to is violation of fundamental alignment requirements [(basic.align#1)](https://timsong-cpp.github.io/cppwp/n4659/basic.align#1): `int32_t` can only be allocated on a four byte boundary and it is known at compile time that `reinterpret_cast<int32_t*>(scratch_space + offset/*=1*/)` is not so aligned.

Reply:
Aha, got it. Thanks for the explanation, and the review of course. Appreciate it!